---
title: "DEG_post_ann"
output: html_document
date: "2025-07-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
library(arrow)
library(Banksy)
library(BiocParallel)
library(DropletUtils)
library(dplyr)
library(edgeR)
library(ggplot2)
library(graphics)
library(igraph)
library(jsonlite)
library(Matrix)
library(matrixStats)
library(OSTA.data)
library(patchwork)
library(pheatmap)
library(S4Vectors)
library(scDblFinder)
library(scater)
library(scran)
library(scuttle)
library(SingleCellExperiment)
library(SpatialExperiment)
library(SummarizedExperiment)
library(tidyr)
library(VisiumIO)
library(spacexr)
library(SpotSweeper)
library(ggspavis)
library(openxlsx)
```

#Confronto: cachetici vs sham (baseline)
#dentro ogni tipo cellulare
#e tra i mionuclei trim dei c26 e i mionuclei 2b degli sham

#oggetto spatial unico (NON RIPETERE)
```{r}

load("~/subset_list.RData")

#per distinguere poi i vari campioni
for (nm in names(subset_list)) {
  spe <- subset_list[[nm]]
  colData(spe)$sample_id <- rep(nm, ncol(spe))  
  subset_list[[nm]] <- spe
}

spe_list <- subset_list[names(subset_list)]
combined_spe <- do.call(cbind, spe_list)
table(colData(combined_spe)$cell_type)
table(colData(combined_spe)$sample_id)

rm(subset_list)

```

#pseudobulk (NON RIPETERE)
```{r}
#il comando di elena per filtrare i geni è troppo pesante e facendo in un altro modo
#me ne rimangono 49. 

# Create pseudo-bulk for each cell_type in each sample
pb_mult <- aggregateAcrossCells(combined_spe, use.assay.type = "counts", 
                                id=DataFrame(label = combined_spe$cell_type,
                                             sample = combined_spe$sample_id))
pb_mult <- logNormCounts(pb_mult)
colnames(pb_mult) <- paste(pb_mult$cell_type, pb_mult$sample_id, sep = "_")
saveRDS(pb_mult, file = "pb_celltype.rds")

```

#DIFFERENTIAL EXPRESSION ANALYSIS
Ho tenuto la normalizzazione TMM di default perché con la upperquantile
molti quantili venivano zero -> errore.
```{r}

pb_mult <- readRDS("~/pb_celltype.rds")

res_mult <- list()
summary_list <- list()

for (cell_type in unique(pb_mult$cell_type)) {
  ct_pb <- pb_mult[, pb_mult$cell_type == cell_type]
  
  y <- DGEList(counts(ct_pb), samples = as.data.frame(colData(ct_pb)))
  
  keep <- filterByExpr(y, group = y$samples$tissue_type)
  y <- y[keep,, keep.lib.sizes=FALSE]
  
  y <- calcNormFactors(y) 
  
  design <- model.matrix(~ 0 + tissue_type, y$samples)
  contrast <- makeContrasts(tissue_typec26 - tissue_typesham,
                            levels = design)
  
  y <- estimateDisp(y, design)
  fit <- glmFit(y, design)
  res_mult[[cell_type]] <- glmLRT(fit, contrast = contrast)
  
  # Risultati
  res <- topTags(res_mult[[cell_type]], n = Inf)$table
  res$significant <- res$FDR < 0.05
  sig <- res[res$significant, ]
  sig_up <- sig[sig$logFC >= 0, ]
  sig_down <- sig[sig$logFC < 0, ]
  
  # Salvataggio in file Excel
  # wb <- createWorkbook()
  # addWorksheet(wb, "Tutti_geni")
  # writeData(wb, "Tutti_geni", res, rowNames = TRUE)
  # addWorksheet(wb, "Significativi")
  # writeData(wb, "Significativi", sig, rowNames = TRUE)
  # addWorksheet(wb, "Upregolati")
  # writeData(wb, "Upregolati", sig_up, rowNames = TRUE)
  # addWorksheet(wb, "Downregolati")
  # writeData(wb, "Downregolati", sig_down, rowNames = TRUE)
  # 
  # clean_name <- gsub("[^[:alnum:]_]", "_", cell_type)
  # saveWorkbook(wb, paste0("dge_", clean_name, ".xlsx"), overwrite = TRUE)
  # 
  # Riepilogo
  gene_summary <- data.frame(
    cell_type = cell_type,
    Geni_considerati = nrow(res),
    Geni_significativi = nrow(sig),
    Upregolati = nrow(sig_up),
    Downregolati = nrow(sig_down)
  )
  
  summary_list[[cell_type]] <- gene_summary
}

#visualizza tabella riassuntiva
summary_df <- do.call(rbind, summary_list)
print(summary_df)

```



#DEGs FOR ALL SAMPLE TYPES 
filterByExpr: roughly speaking, the strategy keeps genes that have at least min.count reads in a worthwhile number of samples.
```{r}

pb_mult <- readRDS("~/pb_celltype.rds")

lrt_list <- list()
summary_list <- list()

contrast_list <- list(
  "C26_vs_SHAM"   = c("tissue_typec26 - tissue_typesham"),
  "FOXO_vs_C26"   = c("-tissue_typec26"),
  "FOXO_vs_SHAM"  = c("-tissue_typesham"),
  "STAT3_vs_C26"  = c("tissue_typec26STAT3 - tissue_typec26"),
  "STAT3_vs_SHAM" = c("tissue_typec26STAT3 - tissue_typesham"),
  "MURF1_vs_C26"  = c("tissue_typec26murf1 - tissue_typec26"),
  "MURF1_vs_SHAM" = c("tissue_typec26murf1 - tissue_typesham"),
  "SMAD23_vs_C26" = c("tissue_typec26SMAD23 - tissue_typec26"),
  "SMAD23_vs_SHAM"= c("tissue_typec26SMAD23 - tissue_typesham")
)

for (contrast_name in names(contrast_list)) {
  lrt_list[[contrast_name]] <- list()
  
  for (cell_type in unique(pb_mult$cell_type)) {
    ct_pb <- pb_mult[, pb_mult$cell_type == cell_type]
    y <- DGEList(counts(ct_pb), samples = as.data.frame(colData(ct_pb)))
    keep <- filterByExpr(y, group = y$samples$tissue_type)
    y <- y[keep,, keep.lib.sizes=FALSE]
    y <- calcNormFactors(y, method="TMM") #è il metodo di default cmq
    design <- model.matrix(~ 0 + tissue_type, y$samples)
    y <- estimateDisp(y, design)
    fit <- glmFit(y, design)
    
    contrast <- makeContrasts(contrasts = contrast_list[[contrast_name]], levels = design)
    
    lrt <- glmLRT(fit, contrast = contrast)
    res <- topTags(lrt, n = Inf)$table
    res$significant <- res$FDR < 0.05
    
    sig <- res[res$significant, ]
    sig_up <- sig[sig$logFC >= 0, ]
    sig_down <- sig[sig$logFC < 0, ]
    
    lrt_list[[contrast_name]][[cell_type]] <- list(
      full = res,
      sig = sig,
      up = sig_up,
      down = sig_down
    )
    
    gene_summary <- data.frame(
      contrast = contrast_name,
      cell_type = cell_type,
      Geni_considerati = nrow(res),
      Geni_significativi = nrow(sig),
      Upregolati = nrow(sig_up),
      Downregolati = nrow(sig_down)
    )
    
    summary_list[[length(summary_list) + 1]] <- gene_summary
  }
}

# Tabella riassuntiva
summary_df <- do.call(rbind, summary_list)
print(summary_df)

```

#salvataggio 
```{r}
library(openxlsx)

outdir <- "results_excel"
if (!dir.exists(outdir)) dir.create(outdir)

for (contrast_name in names(lrt_list)) {
  wb <- createWorkbook()
  
  for (cell_type in names(lrt_list[[contrast_name]])) {
    res     <- lrt_list[[contrast_name]][[cell_type]]$full
    sig     <- lrt_list[[contrast_name]][[cell_type]]$sig
    sig_up  <- lrt_list[[contrast_name]][[cell_type]]$up
    sig_down<- lrt_list[[contrast_name]][[cell_type]]$down
    
    # nome foglio pulito
    clean_name <- gsub("[^[:alnum:]_]", "_", cell_type)
    
    # foglio per cell_type con sottosezioni
    addWorksheet(wb, paste0(clean_name, "_all"))
    writeData(wb, paste0(clean_name, "_all"), res, rowNames = TRUE)
    
    addWorksheet(wb, paste0(clean_name, "_sig"))
    writeData(wb, paste0(clean_name, "_sig"), sig, rowNames = TRUE)
    
    addWorksheet(wb, paste0(clean_name, "_up"))
    writeData(wb, paste0(clean_name, "_up"), sig_up, rowNames = TRUE)
    
    addWorksheet(wb, paste0(clean_name, "_down"))
    writeData(wb, paste0(clean_name, "_down"), sig_down, rowNames = TRUE)
  }
  
  # salva il file excel del contrasto
  clean_contrast <- gsub("[^[:alnum:]_]", "_", contrast_name)
  saveWorkbook(wb, file.path(outdir, paste0("dge_", clean_contrast, ".xlsx")), overwrite = TRUE)
}

```


#Confronto tra mionuclei trim63 dei c26 e mionuclei 2b degli sham
```{r}
pb_mult$group <- paste0(pb_mult$cell_type, "_", pb_mult$tissue_type)
pb_sub <- pb_mult[, pb_mult$group %in% c("Myonuclei_Trim63_c26", "Myonuclei_IIb_sham")]
pb_sub$group <- factor(pb_sub$group, levels = c("Myonuclei_IIb_sham", "Myonuclei_Trim63_c26"))

y <- DGEList(counts(pb_sub), samples = as.data.frame(colData(pb_sub)))
keep <- filterByExpr(y, group = y$samples$group)
y <- y[keep,, keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
design <- model.matrix(~ group, data=y$samples)
y <- estimateDisp(y, design)
fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef = 2)
res <- topTags(lrt, n = Inf)$table
res$significant <- res$FDR < 0.05
sig <- res[res$significant, ]
sig_up <- sig[sig$logFC >= 0, ]
sig_down <- sig[sig$logFC < 0, ]
  
wb <- createWorkbook()
addWorksheet(wb, "Tutti_geni")
writeData(wb, "Tutti_geni", res, rowNames = TRUE)
addWorksheet(wb, "Significativi")
writeData(wb, "Significativi", sig, rowNames = TRUE)
addWorksheet(wb, "Upregolati")
writeData(wb, "Upregolati", sig_up, rowNames = TRUE)
addWorksheet(wb, "Downregolati")
writeData(wb, "Downregolati", sig_down, rowNames = TRUE)
saveWorkbook(wb,"trimc26_2bsham.xlsx", overwrite = TRUE)
  
gene_summary <- data.frame(
    Geni_considerati = nrow(res),
    Geni_significativi = nrow(sig),
    Upregolati = nrow(sig_up),
    Downregolati = nrow(sig_down)
)
  
gene_summary
  
```



