---
title: "DEG_post_ann"
output: html_document
date: "2025-07-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
library(arrow)
library(Banksy)
library(BiocParallel)
library(DropletUtils)
library(dplyr)
library(edgeR)
library(ggplot2)
library(graphics)
library(igraph)
library(jsonlite)
library(Matrix)
library(matrixStats)
library(OSTA.data)
library(patchwork)
library(pheatmap)
library(S4Vectors)
library(scDblFinder)
library(scater)
library(scran)
library(scuttle)
library(SingleCellExperiment)
library(SpatialExperiment)
library(SummarizedExperiment)
library(tidyr)
library(VisiumIO)
library(spacexr)
library(SpotSweeper)
library(ggspavis)
library(openxlsx)
```

Differentially Expressed Genes analysis: comparing gene expression in the same cell type,
or in 2 different cell types across conditions.

#Creating a unique spatial experiment object with all samples (excluding block 7).
```{r}
load("~/subset_list.RData")

#sample identifiers
for (nm in names(subset_list)) {
  spe <- subset_list[[nm]]
  colData(spe)$sample_id <- rep(nm, ncol(spe))
  subset_list[[nm]] <- spe
}

spe_list <- subset_list[names(subset_list)]
exclude_names <- c("sham_b7", "c26STAT3_b7", "c26_b7")
combined_spe <- do.call(
  cbind,
  spe_list[!(names(spe_list) %in% exclude_names)]
)
table(colData(combined_spe)$sample_id)

rm(subset_list)
# Create pseudo-bulk for each cell_type in each sample
pb_without_b7 <- aggregateAcrossCells(combined_spe, use.assay.type = "counts",
                                id=DataFrame(label = combined_spe$cell_type,
                                             sample = combined_spe$sample_id))
pb_without_b7  <- logNormCounts(pb_without_b7) #altre normalizzazioni?
colnames(pb_without_b7 ) <- paste(pb_without_b7$cell_type, pb_without_b7$sample_id, sep = "_")
saveRDS(pb_without_b7 , file = "pb_wb7.rds")


```

#DIFFERENTIAL EXPRESSION ANALYSIS for all contrasts (WITHIN CELL TYPE).
Note: I kept TMM normalization cause upper-quantile did not work due to the high number of zeros.
filterByExpr: roughly speaking, the strategy keeps genes that have at least min.count
reads in a worthwhile number of samples.
```{r}
pb_mult <- readRDS("~/pb_wb7.rds")

lrt_list <- list()
summary_list <- list()

contrast_list <- list(
  "C26_vs_SHAM"   = c("tissue_typec26 - tissue_typesham"),
  "FOXO_vs_C26"   = c("tissue_typec26foxO - tissue_typec26"),
  "FOXO_vs_SHAM"  = c("tissue_typec26foxO - tissue_typesham"),
  "STAT3_vs_C26"  = c("tissue_typec26STAT3 - tissue_typec26"),
  "STAT3_vs_SHAM" = c("tissue_typec26STAT3 - tissue_typesham"),
  "MURF1_vs_C26"  = c("tissue_typec26murf1 - tissue_typec26"),
  "MURF1_vs_SHAM" = c("tissue_typec26murf1 - tissue_typesham"),
  "SMAD23_vs_C26" = c("tissue_typec26SMAD23 - tissue_typec26"),
  "SMAD23_vs_SHAM"= c("tissue_typec26SMAD23 - tissue_typesham")
)

for (contrast_name in names(contrast_list)) {
  lrt_list[[contrast_name]] <- list()
  
  for (cell_type in unique(pb_mult$cell_type)) {
    ct_pb <- pb_mult[, pb_mult$cell_type == cell_type]
    y <- DGEList(counts(ct_pb), samples = as.data.frame(colData(ct_pb)))
    keep <- filterByExpr(y, group = y$samples$tissue_type)
    y <- y[keep,, keep.lib.sizes=FALSE]
    y <- calcNormFactors(y, method="TMM") #Ã¨ il metodo di default cmq
    design <- model.matrix(~ 0 + tissue_type, y$samples)
    y <- estimateDisp(y, design)
    fit <- glmFit(y, design)
    
    contrast <- makeContrasts(contrasts = contrast_list[[contrast_name]], levels = design)
    
    lrt <- glmLRT(fit, contrast = contrast)
    res <- topTags(lrt, n = Inf)$table
    res$significant <- res$FDR < 0.05
    
    sig <- res[res$significant, ]
    sig_up <- sig[sig$logFC >= 0, ]
    sig_down <- sig[sig$logFC < 0, ]
    
    lrt_list[[contrast_name]][[cell_type]] <- list(
      full = res,
      sig = sig,
      up = sig_up,
      down = sig_down
    )
    
    gene_summary <- data.frame(
      contrast = contrast_name,
      cell_type = cell_type,
      Geni_considerati = nrow(res),
      Geni_significativi = nrow(sig),
      Upregolati = nrow(sig_up),
      Downregolati = nrow(sig_down)
    )
    
    summary_list[[length(summary_list) + 1]] <- gene_summary
  }
}

#Summary
summary_df <- do.call(rbind, summary_list)
print(summary_df)

```

#saving results
```{r}
library(openxlsx)

outdir <- "results_excel"
if (!dir.exists(outdir)) dir.create(outdir)

for (contrast_name in names(lrt_list)) {
  wb <- createWorkbook()

  for (cell_type in names(lrt_list[[contrast_name]])) {
    res     <- lrt_list[[contrast_name]][[cell_type]]$full
    sig     <- lrt_list[[contrast_name]][[cell_type]]$sig
    sig_up  <- lrt_list[[contrast_name]][[cell_type]]$up
    sig_down<- lrt_list[[contrast_name]][[cell_type]]$down

    clean_name <- gsub("[^[:alnum:]_]", "_", cell_type)

    addWorksheet(wb, paste0(clean_name, "_all"))
    writeData(wb, paste0(clean_name, "_all"), res, rowNames = TRUE)

    addWorksheet(wb, paste0(clean_name, "_sig"))
    writeData(wb, paste0(clean_name, "_sig"), sig, rowNames = TRUE)

    addWorksheet(wb, paste0(clean_name, "_up"))
    writeData(wb, paste0(clean_name, "_up"), sig_up, rowNames = TRUE)

    addWorksheet(wb, paste0(clean_name, "_down"))
    writeData(wb, paste0(clean_name, "_down"), sig_down, rowNames = TRUE)
  }

  # salva il file excel del contrasto
  clean_contrast <- gsub("[^[:alnum:]_]", "_", contrast_name)
  saveWorkbook(wb, file.path(outdir, paste0("dge_", clean_contrast, ".xlsx")), overwrite = TRUE)
}

```


#Comparison between trim63 myonuclei from c26 samples and myonuclei 2b from sham samples.
```{r}
pb_mult <- readRDS("~/pb_wb7.rds")
pb_mult$group <- paste0(pb_mult$cell_type, "_", pb_mult$tissue_type)
pb_sub <- pb_mult[, pb_mult$group %in% c("Myonuclei_Trim63_c26", "Myonuclei_IIb_sham")]
pb_sub$group <- factor(pb_sub$group, levels = c("Myonuclei_IIb_sham", "Myonuclei_Trim63_c26"))

y <- DGEList(counts(pb_sub), samples = as.data.frame(colData(pb_sub)))
keep <- filterByExpr(y, group = y$samples$group)
y <- y[keep,, keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
design <- model.matrix(~ group, data=y$samples)
y <- estimateDisp(y, design)
fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef = 2)
res <- topTags(lrt, n = Inf)$table
res$significant <- res$FDR < 0.05
sig <- res[res$significant, ]
sig_up <- sig[sig$logFC >= 0, ]
sig_down <- sig[sig$logFC < 0, ]
  
# wb <- createWorkbook()
# addWorksheet(wb, "Tutti_geni")
# writeData(wb, "Tutti_geni", res, rowNames = TRUE)
# addWorksheet(wb, "Significativi")
# writeData(wb, "Significativi", sig, rowNames = TRUE)
# addWorksheet(wb, "Upregolati")
# writeData(wb, "Upregolati", sig_up, rowNames = TRUE)
# addWorksheet(wb, "Downregolati")
# writeData(wb, "Downregolati", sig_down, rowNames = TRUE)
# saveWorkbook(wb,"trimc26_2bsham.xlsx", overwrite = TRUE)

gene_summary <- data.frame(
    Geni_considerati = nrow(res),
    Geni_significativi = nrow(sig),
    Upregolati = nrow(sig_up),
    Downregolati = nrow(sig_down)
)
  
gene_summary
  
```
#3. TRIM C26 VS MYONUCLEI OVERALL SHAM ------------------------------------------------------------------------
```{r}
pb_mult <- readRDS("~/pb_wb7.rds")
pb_mult$group <- paste0(pb_mult$cell_type, "_", pb_mult$tissue_type)
mio_ov_sham <- unique(grep(paste0("^", "Myonuclei", ".*", "sham"), pb_mult$group, value = TRUE))
pb_sub <- pb_mult[, pb_mult$group %in% c("Myonuclei_Trim63_c26", mio_ov_sham)]
pb_sub$group[pb_sub$group %in% mio_ov_sham] <- "Myonuclei_sham"
pb_sub$group <- as.factor(pb_sub$group)
pb_comb <- aggregateData(pb_sub, assay = "counts", fun = "sum", by = "sample_id")
colData(pb_comb)$group <- pb_sub$group[match(colnames(pb_comb), pb_sub$sample_id)]
names(assays(pb_comb)) <- "counts"

y <- DGEList(counts(pb_comb), samples = as.data.frame(colData(pb_comb)))
keep <- filterByExpr(y, group = y$samples$group)
y <- y[keep,, keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
design <- model.matrix(~ group, data=y$samples)
y <- estimateDisp(y, design)
fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef = 2)
top <- topTags(lrt, n = Inf)$table
top$SYMBOL <- rownames(top)

top$significant <- top$FDR < 0.05
sig <- top[top$significant, ]
sig_up <- sig[sig$logFC >= 0, ]
sig_down <- sig[sig$logFC < 0, ]
  
wb <- createWorkbook()
addWorksheet(wb, "Tutti_geni")
writeData(wb, "Tutti_geni", top, rowNames = TRUE)
addWorksheet(wb, "Significativi")
writeData(wb, "Significativi", sig, rowNames = TRUE)
addWorksheet(wb, "Upregolati")
writeData(wb, "Upregolati", sig_up, rowNames = TRUE)
addWorksheet(wb, "Downregolati")
writeData(wb, "Downregolati", sig_down, rowNames = TRUE)
saveWorkbook(wb,"trimc26_overallsham.xlsx", overwrite = TRUE)

gene_summary <- data.frame(
    Geni_considerati = nrow(top),
    Geni_significativi = nrow(sig),
    Upregolati = nrow(sig_up),
    Downregolati = nrow(sig_down)
)
  
gene_summary
  
```

